torch/csrc/jit/argument_spec.cpp

```diff
@@ -186,6 +200,15 @@ std::vector<TypePtr> ArgumentSpecCreator::getSpecializedTypes(
               arg.requires_grad()));
         }
       } break;
+      case SPECIALIZE_NONTENSOR_OPTIONAL: {
+        auto ot = (*input_stack.back()++)->expect<OptionalType>();
+        auto& arg = spec.at(arg_spec_offset++);
+	if (arg.isNone()) {
+	  result_stack.back().emplace_back(NoneType::get());
```

Comment:
This is the source of the problems with NoneType. Explicit None's are represented as:

```
a : Optional[int] = prim::Constant()
```
in the graph. NoneType is only used as an intermediate in the compiler. Instead of trying to change the input to NoneType. Instead I suggest disconnecting the input that must be none from its uses and replacing it with the appropriate None constant. This will require changing the API here:

ArgumentSpecCreator::specializeTypes(Graph& graph, ...) rather than simply returning new types. This change is fine to make, and will end up helping changes that I may need to do to argspec in the future.

Reply:
This has been the thing I've been missing. Thank you so much!

The remaining not so nice thing here is that I don't know how to specialize Optionals to a Typed-None-Constant inside tuples and objects, because I cannot use `replaceAllUses`.
My tentative conclusion is that I cannot do this currently unless I can add some attribute to the optional type to indicate that it is actually a Typed-None constant.
What could be done is to see which uses are unpacking this, but that's really ugly and probably would remain incomplete.